src/GafferImage/Premultiply.cpp

```diff
@@ -0,0 +1,122 @@
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "Gaffer/Context.h"
+
+#include "GafferImage/Premultiply.h"
+
+using namespace IECore;
+using namespace Gaffer;
+
+namespace GafferImage
+{
+
+IE_CORE_DEFINERUNTIMETYPED( Premultiply );
+
+size_t Premultiply::g_firstPlugIndex = 0;
+
+Premultiply::Premultiply( const std::string &name )
+	:	ChannelDataProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new StringPlug( "byChannel", Gaffer::Plug::In, "A" ) );
+}
+
+Premultiply::~Premultiply()
+{
+}
+
+Gaffer::StringPlug *Premultiply::byChannelPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *Premultiply::byChannelPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+void Premultiply::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ChannelDataProcessor::affects( input, outputs );
+
+	if( input == inPlug()->channelDataPlug() ||
+	    input == byChannelPlug() )
+	{
+		outputs.push_back( outPlug()->channelDataPlug() );
+	}
+
+}
+
+void Premultiply::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ChannelDataProcessor::hashChannelData( output, context, h );
+
+	inPlug()->channelDataPlug()->hash( h );
+	byChannelPlug()->hash( h );
+}
+
+void Premultiply::processChannelData( const Gaffer::Context *context, const ImagePlug *parent, const std::string &channel, FloatVectorDataPtr outData ) const
+{
+	if ( channel == byChannelPlug()->getValue() ) return;
+
+	ConstStringVectorDataPtr inChannelsData = inPlug()->channelNamesPlug()->getValue();
+	const std::vector<std::string> &inChannels = inChannelsData->readable();
+
+	if ( std::find( inChannels.begin(), inChannels.end(), byChannelPlug()->getValue() ) == inChannels.end() )
```

Comment:
You mean the error flagging?

If I take that out, it will just return black - that was the original method, but I actually thought it was worth flagging as an error if the channel didn't exist.

If you'd prefer to just return black, I can certainly do that.